client/webserver/webserver.go

```diff
@@ -0,0 +1,402 @@
+	s.mtx.Lock()
+	s.authToken = token
+	s.mtx.Unlock()
+	return token
+}
+
+// token returns the current auth token.
+func (s *WebServer) token() string {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return s.authToken
+}
+
+// watchMarket watches the specified market, cancelling any other market
+// subscriptions.
+func (s *WebServer) watchMarket(dex string, base, quote uint32) (*core.OrderBook, *marketSyncer, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	mktID := marketID(base, quote)
+
+	book, updates, err := s.core.Sync(dex, base, quote)
+	if err != nil {
+		return nil, nil, err
+	}
+	s.watchers[mktID]++
+	ctx, shutdown := context.WithCancel(s.ctx)
+	go func() {
+		log.Debugf("monitoring market %d-%d @ %s", base, quote, dex)
+	out:
+		for {
+			select {
+			case update := <-updates:
+				log.Tracef("order book update received for " + update.Market)
+			case <-ctx.Done():
+				break out
+			}
+		}
+		// Decrement the watchers, and unsync the book if there are none left.
+		s.mtx.Lock()
+		defer s.mtx.Unlock()
+		s.watchers[mktID]--
+		if s.watchers[mktID] == 0 {
+			s.core.Unsync(dex, base, quote)
+		}
+	}()
+	syncer := &marketSyncer{
+		dex:   dex,
+		base:  base,
+		quote: quote,
+		kill:  shutdown,
+	}
+	return book, syncer, nil
+}
+
+// readPost unmarshals the request body into the provided interface.
+func readPost(w http.ResponseWriter, r *http.Request, thing interface{}) bool {
+	body, err := ioutil.ReadAll(r.Body)
+	r.Body.Close()
+	if err != nil {
+		log.Debugf("Error reading request body: %v", err)
+		http.Error(w, "error reading JSON message", http.StatusBadRequest)
+		return false
+	}
+	err = json.Unmarshal(body, &thing)
+	if err != nil {
+		log.Debugf("failed to unmarshal JSON request: %v", err)
+		http.Error(w, "failed to unmarshal JSON request", http.StatusBadRequest)
+		return false
+	}
+	return true
+}
+
+// userInfo is information about the connected user. Some fields are exported
+// for template use.
+type userInfo struct {
+	authed   bool
+	DarkMode bool
+}
+
+// Extract the userInfo from the request context.
+func extractUserInfo(r *http.Request) *userInfo {
+	ai, ok := r.Context().Value(authCV).(*userInfo)
+	if !ok {
+		log.Errorf("no auth info retrieved from client")
+		return &userInfo{}
+	}
+	return ai
+}
+
+// authMiddleware checks incoming requests for cookie-based information
+// including the auth token.
+func (s *WebServer) authMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var authToken string
+		cookie, err := r.Cookie(authCK)
+		switch err {
+		case nil:
+			authToken = cookie.Value
+		case http.ErrNoCookie:
+		default:
+			log.Errorf("authToken retrieval error: %v", err)
+		}
+		authed := authToken != "" && authToken == s.token()
+
+		// darkMode cookie.
+		var darkMode bool
+		cookie, err = r.Cookie(darkModeCK)
+		switch err {
+		// Dark mode is the default
+		case nil, http.ErrNoCookie:
+			darkMode = cookie.Value == "1"
```

Comment:
Seems like this is only good for `case nil:` since `http.ErrNoCookie` would be a `nil` `cookie` and thus a panic.

Also, the comment says "Dark mode is the default", but it's initialized to false.